File: application/app/src/main/java/com/yahyeet/boardbook/presenter/matchfeed/MatchfeedAdapter.java
package com.yahyeet.boardbook.presenter.matchfeed;

import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.yahyeet.boardbook.R;
import com.yahyeet.boardbook.activity.home.matchfeed.matchdetail.MatchDetailActivity;
import com.yahyeet.boardbook.model.entity.Match;
import com.yahyeet.boardbook.model.entity.MatchPlayer;
import com.yahyeet.boardbook.model.entity.User;
import com.yahyeet.boardbook.model.util.StatisticsUtil;
import com.yahyeet.boardbook.presenter.BoardbookSingleton;

import java.util.ArrayList;
import java.util.List;


/**
 * Adapter for the match feed recycler view
 */
public class MatchfeedAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

	private static final String TAG = "MatchfeedAdapter";
	private static final int PROFILE_HEADER_VIEW = 1;
	private Context context;
	private List<Match> matches;
	private User currentUser;
	private boolean isProfile = false;

	private StatisticsUtil statisticsUtil;


	public MatchfeedAdapter(Context context, List<Match> dataset, User currentUser) {
		if (dataset != null)
			matches = dataset;
		else
			matches = new ArrayList<>();
		this.currentUser = currentUser;

		this.context = context;
	}

	public MatchfeedAdapter(Context context, List<Match> dataset, User user, StatisticsUtil statisticsUtil) {
		if (dataset != null)
			matches = dataset;
		else
			matches = new ArrayList<>();
		currentUser = user;
		this.context = context;
		this.statisticsUtil = statisticsUtil;

		if(currentUser != null)
			isProfile = true;
	}


	public class HeaderViewHolder extends RecyclerView.ViewHolder {
		private TextView tvUsername;
		private TextView tvWinrate;
		private TextView tvGamesPlayed;
		private ProgressBar pbWinrate;

		HeaderViewHolder(@NonNull View v) {
			super(v);
			//Does nothing when clicked on!

			tvUsername = v.findViewById(R.id.tvUsername);
			tvWinrate = v.findViewById(R.id.tvWinrate);
			tvGamesPlayed = v.findViewById(R.id.tvGamesPlayed);
			pbWinrate = v.findViewById(R.id.pbWinrate);
		}
	}

	public class MatchViewHolder extends RecyclerView.ViewHolder {

		private TextView tvWinLost;
		private TextView tvGameName;
		private TextView tvRoleName;
		private TextView tvTeamName;
		private TextView tvPlayerAmount;
		private ImageView imageView;

		MatchViewHolder(View v) {
			super(v);
			// Define click listener for the ViewHolder's View.
			v.setOnClickListener(v1 -> Log.d(TAG, "Element " + getAdapterPosition() + " clicked."));

			tvWinLost = v.findViewById(R.id.matchWinLost);
			tvGameName = v.findViewById(R.id.matchGameName);
			tvRoleName = v.findViewById(R.id.matchRoleName);
			tvTeamName  = v.findViewById(R.id.matchTeamName);
			tvPlayerAmount = v.findViewById(R.id.matchPlayerAmount);
			imageView = v.findViewById(R.id.matchGameImage);
		}
	}


	// Creates new view, does not assign data
	@NonNull
	@Override
	public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
		View v;

		if (viewType == PROFILE_HEADER_VIEW) {
			v = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.element_profile_header, viewGroup, false);
			return new HeaderViewHolder(v);
		}
		v = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.element_match, viewGroup, false);
		return new MatchViewHolder(v);
	}

	// Replace the contents of a view (invoked by the layout manager)
	// Method that assigns data to the view
	@Override
	public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {


		if(isProfile)
			// Position - 1 to align match list with position,
			position -= 1;

		try {
			if (holder instanceof MatchViewHolder) {
				// For onClickListener lambda expression
				int finalPosition = position;

				MatchPlayer currentMatchPlayer = matches
					.get(position)
					.getMatchPlayerByUser(currentUser);

				MatchViewHolder vh = (MatchViewHolder) holder;
				vh.tvWinLost.setText(currentMatchPlayer.getWin() ? "Winner" : "Looser");
				vh.tvGameName.setText(matches.get(position).getGame().getName());
				vh.tvPlayerAmount.setText(matches.get(position).getMatchPlayers().size() + " Players");

				if(currentMatchPlayer.getTeam() != null){
					vh.tvTeamName.setText("In " + currentMatchPlayer.getTeam().getName());
					if(currentMatchPlayer.getRole() != null)
						vh.tvRoleName.setText("(" + currentMatchPlayer.getRole().getName() + ")");
					else
						vh.tvRoleName.setText("");
				} else if(currentMatchPlayer.getRole() != null){
					vh.tvTeamName.setText("as " + currentMatchPlayer.getRole().getName());
					vh.tvRoleName.setText("");
				}
				else{
					vh.tvRoleName.setText("");
					vh.tvTeamName.setText("");
				}


				vh.itemView.setOnClickListener(view -> {
					Intent intent = new Intent(context, MatchDetailActivity.class);
					intent.putExtra("Match", matches.get(finalPosition).getId());
					context.startActivity(intent);
				});

				//vh.imageView.setImageURL();
			} else if (holder instanceof HeaderViewHolder) {
				HeaderViewHolder vh = (HeaderViewHolder) holder;
				vh.tvUsername.setText(currentUser.getName());
				double stats = statisticsUtil.getWinrateFromMatches(currentUser.getMatches(), currentUser);
				int percent = (int) (100 * stats);
				vh.tvGamesPlayed.setText(Integer.toString(currentUser.getMatches().size()));
				vh.tvWinrate.setText(percent + "%");
				vh.pbWinrate.setProgress(percent);
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	@Override
	public int getItemCount() {
		if (matches == null) {
			return 0;
		}

		if(!isProfile)
			return matches.size();

		return matches.size() + 1;
	}

	@Override
	public int getItemViewType(int position) {
		if (position == 0 && isProfile) {
			return PROFILE_HEADER_VIEW;
		}
		return super.getItemViewType(position);
	}
}

File: application/app/src/main/java/com/yahyeet/boardbook/presenter/matchfeed/MatchfeedAdapter.kt
package com.yahyeet.boardbook.presenter.matchfeed

import android.content.Context
import android.content.Intent
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

import com.yahyeet.boardbook.R
import com.yahyeet.boardbook.activity.home.matchfeed.matchdetail.MatchDetailActivity
import com.yahyeet.boardbook.model.entity.Match
import com.yahyeet.boardbook.model.entity.MatchPlayer
import com.yahyeet.boardbook.model.entity.User
import com.yahyeet.boardbook.model.util.StatisticsUtil
import com.yahyeet.boardbook.presenter.BoardbookSingleton

import java.util.ArrayList


class MatchfeedAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private var context: Context? = null
    private var matches: List<Match>? = null

    //TODO: Naming might need to change
    private val user: User?
    private val statisticsUtil: StatisticsUtil


    constructor(context: Context, dataset: List<Match>?) {
        if (dataset != null)
            matches = dataset
        else
            matches = ArrayList()

        this.context = context
    }

    constructor(context: Context, dataset: List<Match>?, user: User, statisticsUtil: StatisticsUtil) {
        if (dataset != null)
            matches = dataset
        else
            matches = ArrayList()

        this.context = context
        this.user = user
        this.statisticsUtil = statisticsUtil
    }


    inner class HeaderViewHolder internal constructor(v: View) : RecyclerView.ViewHolder(v) {
        private val tvUsername: TextView
        private val tvWinrate: TextView
        private val tvGamesPlayed: TextView
        private val pbWinrate: ProgressBar

        init {
            //Does nothing when clicked on!

            tvUsername = v.findViewById(R.id.tvUsername)
            tvWinrate = v.findViewById(R.id.tvWinrate)
            tvGamesPlayed = v.findViewById(R.id.tvGamesPlayed)
            pbWinrate = v.findViewById(R.id.pbWinrate)
        }
    }

    inner class MatchViewHolder internal constructor(v: View) : RecyclerView.ViewHolder(v) {

        // TODO Replace this area with match class as a custom view object
        private val tvWinLost: TextView
        private val tvGameName: TextView
        private val tvRoleName: TextView
        private val tvTeamName: TextView
        private val tvPlayerAmount: TextView
        private val imageView: ImageView


        init {
            // Define click listener for the ViewHolder's View.
            v.setOnClickListener { v1 -> Log.d(TAG, "Element $adapterPosition clicked.") }

            tvWinLost = v.findViewById(R.id.matchWinLost)
            tvGameName = v.findViewById(R.id.matchGameName)
            tvRoleName = v.findViewById(R.id.matchRoleName)
            tvTeamName = v.findViewById(R.id.matchTeamName)
            tvPlayerAmount = v.findViewById(R.id.matchPlayerAmount)
            imageView = v.findViewById(R.id.matchGameImage)
        }
    }


    // Creates new view, does not assign data
    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val v: View

        if (viewType == PROFILE_HEADER_VIEW) {
            v = LayoutInflater.from(viewGroup.context).inflate(R.layout.element_profile_header, viewGroup, false)
            return HeaderViewHolder(v)
        }
        v = LayoutInflater.from(viewGroup.context).inflate(R.layout.element_match, viewGroup, false)
        return MatchViewHolder(v)
    }

    // Replace the contents of a view (invoked by the layout manager)
    // Method that assigns data to the view
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        // TODO: Replace with model integration when match is implemented

        try {
            if (holder is MatchViewHolder) {
                val currentMatchPlayer = matches!![position]
                        .getMatchPlayerByUser(BoardbookSingleton.getInstance().authHandler.loggedInUser)

                holder.tvWinLost.text = if (currentMatchPlayer.win) "Winner" else "Looser"
                holder.tvGameName.text = matches!![position].game!!.name
                holder.tvPlayerAmount.text = matches!![position].matchPlayers.size.toString() + " Players"

                if (currentMatchPlayer.team!!.name != null) {
                    holder.tvTeamName.text = "In " + currentMatchPlayer.team!!
                    if (currentMatchPlayer.role!!.name != null)
                        holder.tvRoleName.text = "(" + currentMatchPlayer.role + ")"
                } else if (currentMatchPlayer.team!!.name == null && currentMatchPlayer.role!!.name != null) {
                    holder.tvTeamName.text = "as " + currentMatchPlayer.role!!.name!!
                    holder.tvRoleName.text = ""
                }



                holder.itemView.setOnClickListener { view ->
                    val intent = Intent(context, MatchDetailActivity::class.java)
                    intent.putExtra("Match", matches!![position].id)
                    context!!.startActivity(intent)
                }

                //vh.imageView.setImageURL();
            } else if (holder is HeaderViewHolder) {
                holder.tvUsername.text = user!!.name
                val stats = statisticsUtil.getWinrateFromMatches(matches!!, user)
                val percent = (100 * stats).toInt()
                holder.tvGamesPlayed.text = Integer.toString(user.matches.size)
                holder.tvWinrate.text = "$percent%"
                holder.pbWinrate.progress = percent
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun getItemCount(): Int {
        if (matches == null) {
            return 0
        }

        return if (user == null) matches!!.size else matches!!.size + 1

    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0 && user != null) {
            PROFILE_HEADER_VIEW
        } else super.getItemViewType(position)
    }

    companion object {

        private val TAG = "MatchfeedAdapter"
        private val PROFILE_HEADER_VIEW = 1
    }
}
